File: challenges-in-python/dynamic_programming/knapsack.py
# Solving the "0/1 Knapsack" Challenge
# Challenge Description:
# Given two integer arrays vals (representing the values of the items) 
# and weights (representing the weights of the items), 
# along with an integer W representing the maximum weight capacity of the knapsack, 
# determine the maximum value you can achieve by choosing items such that 
# their total weight is less than or equal to W. 
# Each item can only be selected once (0/1 property).

# Approach:
# A dynamic programming solution involves creating a 2D DP table 
# where dp[i][w] represents the maximum value that can be achieved with the first i items 
# and a maximum weight limit of w.

# The base case is when i = 0 or w = 0,
# where the maximum value that can be achieved is 0.

# The recursive case is when the weight of the current item is less than or equal to w.
# In this case, we have two choices:
# 1. We can include the current item in the knapsack,
#    in which case the maximum value that can be achieved is dp[i-1][w-weights[i]] + vals[i].
# 2. We can exclude the current item from the knapsack,
#    in which case the maximum value that can be achieved is dp[i-1][w].

# The maximum value that can be achieved is the maximum of the two choices.

# If the weight of the current item is greater than w,
# we can only exclude the current item from the knapsack,
# in which case the maximum value that can be achieved is dp[i-1][w].

# The final answer is stored in dp[n][W], where n is the number of items.

# Big O:
# Time complexity: O(nW) | Space complexity: O(nW)
def knapsack(values, weights, W):
    n = len(values)
    dp = [[0] * (W+1) for _ in range(n+1)]
    for i in range(1, n+1):
        for w in range(1, W+1):
            if weights[i-1] <= w:
                dp[i][w] = max(dp[i-1][w-weights[i-1]] + values[i-1], dp[i-1][w])
            else:
                dp[i][w] = dp[i-1][w]

    return dp[n][W]

# Test cases
def test_knapsack():
    values = [60, 100, 120]
    weights = [10, 20, 30]
    W = 50
    assert knapsack(values, weights, W) == 220

test_knapsack()

File: challenges-in-rust/src/dynamic_programming/knapsack.rs
// Solving the "0/1 Knapsack" Challenge
// Problem Statement
// Given two integer arrays vals (representing the values of items) and weights (representing the weights of items), 
// along with an integer W representing the maximum capacity of a knapsack, 
// find the maximum value you can achieve by choosing items such that the total weight does not exceed W. 
// Each item can either be included or excluded (hence 0/1 Knapsack).

// Example
// vals = [60, 100, 120]
// weights = [10, 20, 30]
// W = 50
// Output: 220

// Big O:
// Time cmp: O(n * W) | Space cmp: O(n * W)
pub fn knapsack(vals: Vec<i32>, weights: Vec<i32>, maximum_capacity: i32) -> i32 {
    let n = vals.len();
    // dp[i][w] is the maximum value that can be attained with weight less than or equal to w using the first i items
    // vec![0; (w + 1) as usize]: This part creates a vector of integers initialized with zeros. The size of this vector is w + 1
    // vec![vec![0; (w + 1) as usize]; n + 1]: This part creates a vector of vectors of integers. The size of the outer vector is n + 1. The size of the inner vector is w + 1
    let mut dp = vec![vec![0; maximum_capacity as usize + 1]; n + 1];
    for i in 0..=n {
        for w in 0..=maximum_capacity as usize {
            if i == 0 || w == 0 {
                dp[i][w] = 0;
            } else if weights[i - 1] <= w as i32 {
                dp[i][w] = std::cmp::max(
                    vals[i - 1] + dp[i - 1][w - weights[i - 1] as usize], //
                    dp[i - 1][w],
                );
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }
    dp[n][maximum_capacity as usize]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_knapsack() {
        assert_eq!(knapsack(vec![60, 100, 120], vec![10, 20, 30], 50), 220);
        assert_eq!(knapsack(vec![60, 100, 120], vec![10, 20, 30], 0), 0);
    }
}
